### PR TITLE
fix: don't show warning on numpad

### LINKED ISSUE #17490: "Press Alt+K to show keyboard shortcuts" pops up on Numpad 1
> Basically title. Every time I press Numpad 1 (Again), I get the shortcut reminder. It doesn't happen on keyboard 1 or on-screen press, or any other input [Update: Now happening with Num Pad 1, 2, 3, 4]. I don't think I've made any recent changes, but I think there was a recent update to Ankidroid.
> 
> Edit: I forget how complicated all this. I'm using an external bluetooth keyboard AND running it all through External Keyboard Helper with International Keyboard layout. I have used this setup without any problems previously.

https://www.reddit.com/r/Anki/comments/1gyolng/ankidroid_press_altk_to_show_keyboard_shortcuts/

```
AnkiDroid Version = 2.19.2 (b073f52783583fd4fd1436d986090ced9006a66b)

Backend Version = 0.1.43-anki24.06.3 (24.06.3 d678e39350a2d243242a69f4e22f5192b04398f2)

Android Version = 14 (SDK 34)

ProductFlavor = play

Manufacturer = samsung

Model = SM-X210

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 14; SM-X210 Build/UP1A.231005.007; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/131.0.6778.39 Safari/537.36

ACRA UUID = 0630523f-9578-4f99-9ee3-3fd09cc1e0f4

FSRS = 0.6.4 (Enabled: true)

Crash Reports Enabled = true
```

### PR DESCRIPTION
**Pull request by SanjaySargam**
## Purpose / Description
"Press Alt+K to show keyboard shortcuts" pops up on Numpad 1

## Approach
Previously the code was incorrectly handled which it only checks isNumKey pressed then condition getting true which is not expected. I group isNumKey condition with isModifierKey it will be true when both modifier and num key pressed.

## Fixes
* Fixes #17490

## How Has This Been Tested?
Chromebook 

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by mikehardy** [comment]
@SanjaySargam our PR template has an Approach section typically, where you would describe the *why* behind the change, that can help in review.

I see the change and I can read the code and puzzle it out, but it is much quicker if you use the Approach section to say something like:

"realized that the existing code makes XYZ assumption which isn't always true / mixes two ideas, split it into separate checks to more correctly NNN the AAA" or something

**Review by mikehardy — CHANGES_REQUESTED** [review]
So it looks like you have done a refactor but I don't see any change in functionality

Did you reproduce this?
Can you show a video reproducing it before and not-reproducing it after?

If I'm reading the code correctly (I may not be - it is hard to review things when there is a refactor and a functional change mixed in the same commit) then I don't see anything that would/should affect behavior here

**Review by david-allison — APPROVED** [review]
TL;DR: needs a line in the commit message to explain the bug and why the change fixes it

----


I believe the change is:

```diff
- && (keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9)) {
+ && ((keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9))) {
```

Previously, the "or" at the end incorrectly took priority in the expression, so a numkey WITHOUT a modifier would return `true`

----

In which, code approved, but needs a commit message change or test

**Comment by mikehardy** [comment]
Yeah I dunno, can't make much sense of it with the refactor mixed in 🤷 . I just re-read it, still can't, and with no further explanation 🤷

**Comment by BrayanDSO** [comment]
Tbh, I still don't know why we don't use Android's default shortcut, which is `Meta+/`. "But not all keyboards have a meta key", also use the other suggested Android keys (source https://developer.android.com/develop/ui/compose/touch-input/keyboard-input/keyboard-shortcuts-helper)

```
Note: The Meta key is not present on all keyboards. On macOS keyboards, the Meta key is the Command key; on Windows keyboards, the Windows key; and on ChromeOS keyboards, the Search key.
```

The user may have system shortcuts that AnkiDroid can't recognize for doing other things, so warning every time a unknown shortcut is used isn't sane IMO

**Comment by SanjaySargam** [comment]
@BrayanDSO I already tried this:
```kotlin
        if (event.isMetaPressed && keyCode == KeyEvent.KEYCODE_SLASH) {
            showKeyboardShortcutsDialog()
            return true
        }
``` 
but it's not working.

**Review by david-allison — CHANGES_REQUESTED** [review]
I no longer understand how this works, and am concerned the 'fix' has introduced a regression.

[Is a unit test feasible here]?

**Comment by SanjaySargam** [comment]
@david-allison I clearly mentioned in commit message. Did you refer that?

**Comment by david-allison** [comment]
@SanjaySargam 

I don't see how the patch fixes the problem. 

This test (using the old code) fails, and the test continues to fail with the fix

```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt b/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt	(revision 56aa7f4ff7e068f3cc96f3170b56f68dddf9e0ee)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt	(date 1732899258427)
@@ -19,6 +19,7 @@
 
 import android.view.KeyEvent
 import android.view.KeyboardShortcutInfo
+import androidx.annotation.CheckResult
 import androidx.annotation.StringRes
 import com.ichi2.anki.AnkiActivityProvider
 import com.ichi2.anki.CollectionManager.TR
@@ -76,6 +77,12 @@
             else -> KeyEvent.keyCodeFromString(key)
         }
     }
+
+    companion object {
+        @CheckResult
+        fun isUnmappedShortcut(event: KeyEvent, keyCode: Int) : Boolean =
+            (event.isCtrlPressed || event.isAltPressed || event.isMetaPressed) && (keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9)
+    }
 }
 
 /**
Index: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt	(revision 56aa7f4ff7e068f3cc96f3170b56f68dddf9e0ee)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.kt	(date 1732898448164)
@@ -52,6 +52,7 @@
 import com.ichi2.anim.ActivityTransitionAnimation.Direction.DEFAULT
 import com.ichi2.anim.ActivityTransitionAnimation.Direction.NONE
 import com.ichi2.anki.analytics.UsageAnalytics
+import com.ichi2.anki.android.input.Shortcut
 import com.ichi2.anki.android.input.ShortcutGroup
 import com.ichi2.anki.android.input.ShortcutGroupProvider
 import com.ichi2.anki.android.input.shortcut
@@ -680,8 +681,10 @@
 
         val done = super.onKeyUp(keyCode, event)
 
+        if (done || shortcuts == null) return false
+
         // Show snackbar only if the current activity have shortcuts, a modifier key is pressed and the keyCode is an unmapped alphabet key
-        if (!done && shortcuts != null && ((event.isCtrlPressed || event.isAltPressed || event.isMetaPressed) && (keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9))) {
+        if (Shortcut.isUnmappedShortcut(event, keyCode)) {
             showSnackbar(R.string.show_shortcuts_message, Snackbar.LENGTH_SHORT)
             return true
         }
Index: AnkiDroid/src/test/java/com/ichi2/anki/ShortcutTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/ShortcutTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/ShortcutTest.kt
new file mode 100644
--- /dev/null	(date 1732899237347)
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/ShortcutTest.kt	(date 1732899237347)
@@ -0,0 +1,34 @@
+/*
+ *  Copyright (c) 2024 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki
+
+import android.view.KeyEvent
+import com.ichi2.anki.android.input.Shortcut
+import org.mockito.kotlin.mock
+import kotlin.test.Test
+import kotlin.test.assertFalse
+
+class ShortcutTest {
+    val keyEventWithNoModifiers : KeyEvent = mock { }
+
+    @Test
+    fun `single number is not a shortcut hint`() {
+        assertFalse(
+            Shortcut.isUnmappedShortcut(keyEventWithNoModifiers, KeyEvent.KEYCODE_NUMPAD_1)
+        )
+    }
+}
\ No newline at end of file
```

**Comment by SanjaySargam** [comment]
> I don't see how the patch fixes the problem.

I tested it on Chromebook, it fixed

**Comment by david-allison** [comment]
Could you look into why the unit test (which I believe reproduces the problem) still fails when your patch is applied?

Ideally include a test in your patch, this has been confusing for a number of people.

**Comment by SanjaySargam** [comment]
Really sorry for that. I now I messed up. Will be pushing updated commit

**Comment by david-allison** [comment]
For reference, I would solve this as:


```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt b/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt	(revision 2c2fe1b46e8912eaadbb329d814de824f0ea5b4b)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt	(date 1732899886387)
@@ -80,8 +80,12 @@
 
     companion object {
         @CheckResult
-        fun isUnmappedShortcut(event: KeyEvent, keyCode: Int): Boolean =
-            (event.isCtrlPressed || event.isAltPressed || event.isMetaPressed) && (keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9)
+        fun isUnmappedShortcut(event: KeyEvent, keyCode: Int): Boolean {
+            val modifierPressed = (event.isCtrlPressed || event.isAltPressed || event.isMetaPressed)
+            if (!modifierPressed) return false
+            return (keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || 
+                    (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9)
+        }
     }
 }
```

**Comment by david-allison** [comment]
No worries! Team effort. It's what reviews are for!

I believe the original patch worked. The simplified one didn't

**Comment by SanjaySargam** [comment]
```kotlin
    companion object {
        @CheckResult
        fun isUnmappedShortcut(event: KeyEvent, keyCode: Int): Boolean =
            (event.isCtrlPressed || event.isAltPressed || event.isMetaPressed) && ((keyCode in KeyEvent.KEYCODE_A..KeyEvent.KEYCODE_Z) || (keyCode in KeyEvent.KEYCODE_NUMPAD_0..KeyEvent.KEYCODE_NUMPAD_9))
    }
``` 

I changed like this. And it works/fixed problem. Can we go with this ?

File: AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt [hunk 1]
```diff
@@ -76,6 +77,12 @@ data class Shortcut(val shortcut: String, val label: String) {
             else -> KeyEvent.keyCodeFromString(key)
         }
     }
+
+    companion object {
+        @CheckResult
+        fun isValidShortcutCombination(event: KeyEvent, keyCode: Int): Boolean =
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/android/input/Shortcut.kt:83]: `isPotentialShortcutCombination`

**Review by david-allison — APPROVED** [review]
I'll leave this to Reviewer 2

It fixes the bug, I'd rather the code were a little more clear but I won't insist on it

**Comment by david-allison** [comment]
@SanjaySargam could you rebase this one on `main`, I've put in a potential fix for the test failures

**Review by mikehardy — APPROVED** [review]
I've done the rebase, let's get this in
Now I can see the parentheses moving at least - they're not jumbled in with other changes, and it has test support

**Comment by david-allison** [comment]
Given this is pending merge, maintainers are also free to rebase & force push

**Comment by mikehardy** [comment]
```

CreateDeckDialogTest > searchDecksIconVisibilityDeckCreationTest FAILED

...

        Caused by:
        java.lang.OutOfMemoryError: Java heap space
```

Still having semi-persistent macOS heap space issues but I thought this was rebased to the current main that includes the fix? Perhaps we need to increase heap space ?

**Comment by david-allison** [comment]
I'd support:

* Adding `@JvmTest`
* If that fails, ignore the test in CI
* Adding a follow-up issue to investigate
* getting this merged

**Comment by david-allison** [comment]
image
